### PR TITLE
Separating the protocol from the ledger

### PR DESCRIPTION
**Pull request by JaredCorduan**
This noisy :drum: PR creates an new package `cardano-protocol-tpraos` and moves the following into it:

* Block Header and associated types and helper functions
* Operational Certificate types and helper functions
* Overlay Schedule support
* The following rules: `PRTCL`, `OCERT`, `OVERLAY`, `Updn`

### CODE REVIEW

File: alonzo/impl/cardano-ledger-alonzo.cabal [hunk 1]
```diff
@@ -61,6 +61,7 @@ library
     cardano-ledger-core,
     cardano-ledger-shelley-ma,
     cardano-prelude,
+    cardano-protocol-tpraos,
```
Comment (nc6) [inline alonzo/impl/cardano-ledger-alonzo.cabal:64]: Ultimately I guess we'd like to drop this dependency.

File: cardano-protocol-tpraos/src/Cardano/Protocol/TPraos/BHeader.hs [hunk 2]
```diff
@@ -0,0 +1,490 @@
...
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE PatternSynonyms #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE TypeApplications #-}
+{-# OPTIONS_GHC -Wno-orphans #-}
+
+-- BlockNo is an orphan until the instance is added to cardano-base
```
Comment (nc6) [inline cardano-protocol-tpraos/src/Cardano/Protocol/TPraos/BHeader.hs]: Is this comment in the right place?

File: shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/STS/Chain.hs [hunk 3]
```diff
@@ -53,6 +53,15 @@ import Cardano.Ledger.Serialization (ToCBORGroup)
 import Cardano.Ledger.Shelley.Constraints (UsesValue)
 import Cardano.Ledger.Slot (EpochNo)
 import qualified Cardano.Ledger.Val as Val
+import Cardano.Protocol.TPraos (PoolDistr (..))
+import Cardano.Protocol.TPraos.BHeader (BHeader, LastAppliedBlock (..), bHeaderSize, bhHash, bhbody, bheaderBlockNo, bheaderSlotNo, hBbsize, lastAppliedHash, prevHashToNonce)
```
Comment (nc6) [inline shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/STS/Chain.hs]: Break over multiple lines?

**Review by nc6 — APPROVED** [review]
So I think we'll need to pull things apart further, to remove the dependency at all. But this certainly looks like a solid first step!

File: cardano-protocol-tpraos/src/Cardano/Protocol/TPraos/BHeader.hs (hunk 2 above)
Comment (JaredCorduan) [reply cardano-protocol-tpraos/src/Cardano/Protocol/TPraos/BHeader.hs]: let me see if I can find a home for it where ormolu won't move it...

File: shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/STS/Chain.hs (hunk 3 above)
Comment (JaredCorduan) [reply shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/STS/Chain.hs]: oh wow, I didn't realize how long that one got :laughing:

**Comment by JaredCorduan** [comment]
@nc6 do you think now is an okay time to merge this possibly disruptive PR?

**Comment by nc6** [comment]
Yep, seems reasonable to me!

**Pull request merged**